### ISSUE TITLE
Bypass wake word for an external event such as face recognition

### ISSUE THREAD

**Issue by RobertLukan** [rhasspy/wyoming-satellite#81]
Is there a way to by-pass the wake word and start STT on the requested ? I have face recognition system done with Frigate/Doube-Take/Compreface/Home Assistant. I have connected with Node-red event listening and sending TTS to remote satellite by using Rhasspy. But I noticed that Rhasspy is not developed anymore, so I am trying to migrate to Wyoming satellite.

I have Wyoming satellite running on RPI, I need to make more testing, however I would like to start the conversation with the Home Assistant on successful face recognition and not only on the wake word. 

I was thinking is there a way to send Wyoming satellite HTTP command to start streaming audio for STT and bypass wake command? As I already have Node-red flow ready, I would just need to issue another HTTP call to the satellite.

**Comment by llluis** [comment]
How would you handle the conversation? 

There’s an easy way to implement a remote trigger to the satellite but there’s no context, i.e., it’s exactly the same as you would trigger via wake word and you would have to give complete instructions, not only a reply. 

Are you able to do differently today with rhaspy or node-red?

**Comment by RobertLukan** [comment]
Well, it would be perfect if the context would stay the same. My initial idea was to just replace wake word with an external trigger, simple as that. But I guess the proper way would be to inject data into Home Assistant(Assist) and let the HA or chat bot to answer.

An example:

1. HA event happens(face recognition), it triggers Node-red or Automation based on the event
2. Than automation/Node-red would "say" Mark came home to Chat-bot
3. Chat-bot would respond "Welcome home Mark". I guess this would require to have chat-bot trained(I guess that is doable)
4. Person replies with the wake command and the response. (that is an optional step, person can also just stay silent)

**Comment by llluis** [comment]
OK, I have a working demo. Willing to try it?

You need to copy the contents of this repo (https://github.com/llluis/homeassistant-wyoming) inside your home assistant `/config/custom_components/wyoming` and replace your `satellite.py` **in the RPi** by this one: https://github.com/llluis/wyoming-satellite/blob/remote-detection/wyoming_satellite/satellite.py

Reboot everything and you will have a new button in the satellite device in Home Assistant: 

![image](https://github.com/rhasspy/wyoming-satellite/assets/708868/c119cb3a-7258-4e38-bac2-718a3b47aab7)

You can use this button in automations to trigger the speech to text of the satellite.
Let me know if it works.

**Comment by llluis** [comment]
For the TTS, you can install MPD in the RPi and have this functionality.
See #28 and #40.

**Comment by RobertLukan** [comment]
Fantastic, I will give a try during the weekend. I will study that part of the code and play with it a bit. Now I see which part of the code is involved in this operation.

I actually like how Rhasspy is handling TTS, I wish that we could in similar way do TTS directly by invoking some HA API as part of the Assist(with the context). For now I will follow the MPD way.

**Comment by RobertLukan** [comment]
I just tested it and it works over HA GUI, I just need to figure it out how to do it over HTTP API. 

Do you maybe know if HA Assist have API ? If yes, I could start contextual conversation with Node-red. 

I am facing other issue, like false positive triggering wake word. I will play with this setup a bit more, but it looks buggy to me. I will open other issue once I have more data.

**Comment by RobertLukan** [comment]
Now I know how to get and track conversation_id that is part of the HA API. I am attaching an example from Node-red.

So the next issue is how to associate the conversation id as part of the "wyoming conversation" so that TTS and STT are part of the same conversation. I am not sure what is the correct way forward to be honest.

<img width="1009" alt="Screenshot 2024-01-27 at 20 50 59" src="https://github.com/rhasspy/wyoming-satellite/assets/15019985/5be0acf2-f88f-4017-9f7c-6eb8727e347d">

**Comment by RobertLukan** [comment]
Actually I found this feature within Rhasspy3(fork): https://github.com/rhasspy/rhasspy3/compare/master...Shulyaka:rhasspy3:home_assistant_conversation_id

It is a pity that Rhasspy3 is not being developed anymore. I tried wyoming-satellite and Rhasspy3 and Rhasspy3 works better. Wake word works way better. 

I think I will try more Rhasspy3 for now. Regardless thank you for your help.

**Comment by geman220** [comment]
> OK, I have a working demo. Willing to try it?
> 
> **You must be in `2024.1.5` HA version.** You need to copy the contents of this repo (https://github.com/llluis/homeassistant-wyoming) inside your home assistant `/config/custom_components/wyoming` and replace your `satellite.py` **in the RPi** by this one: https://github.com/llluis/wyoming-satellite/blob/remote-detection/wyoming_satellite/satellite.py
> 
> Reboot everything and you will have a new button in the satellite device in Home Assistant:
> 
> ![image](https://private-user-images.githubusercontent.com/708868/300121113-c119cb3a-7258-4e38-bac2-718a3b47aab7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDcxODAyMDMsIm5iZiI6MTcwNzE3OTkwMywicGF0aCI6Ii83MDg4NjgvMzAwMTIxMTEzLWMxMTljYjNhLTcyNTgtNGUzOC1iYWMyLTcxOGEzYjQ3YWFiNy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMjA2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDIwNlQwMDM4MjNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lY2I3NDY4MDE0ZDhkN2JjMDNjYjg4ZDY1MDBlZGY4OGM5YjI2NDU4NzUyMGE4MjE2MDM2YTgyMzI3MWUzNTdlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.AZvBBMTAI04Q7GMXwNzNcB9FHYN6lWwBRtglPOx-U6o)
> 
> You can use this button in automations to trigger the speech to text of the satellite without the wake word. Let me know if it works.

I don't mean to necro a completed item, so feel free to tell me to open a new item, but I am working to test this.  I need a way to trigger voice via automation, so TTS gets sent to the satellite over MPD, and then I need the satellite to listen for the voice input for a response (yes or no) without the wake word being spoken to determine what to do next.  If you are still looking for someone to test, I can provide feedback as I work through this.

**Comment by RobertLukan** [comment]
Ok, I am willing to test a new version. I got stuck with Rhasspy3 as I dont have time to debug right now.

**Comment by geman220** [comment]
So if I understand correctly, I should be able to “press” on trigger detection in the UI, and that should make the satellite activate (as if a wake word was spoken) and perform the speech to text translation?  Should the assist in progress sensor update to “on”?

**Comment by llluis** [comment]
> So if I understand correctly, I should be able to “press” on trigger detection in the UI, and that should make the satellite activate (as if a wake word was spoken) and perform the speech to text translation? Should the assist in progress sensor update to “on”?

Exactly. However, you will have to give an action to the satellite. It's not possible (yet) to just capture the reply (the transcription). 

Actually, I have a working version with this too, but still a proof of concept.

**Comment by llluis** [comment]
No, the button shall trigger the satellite without any other automation. Did you use my version of the satellite too?

**Comment by geman220** [comment]
> No, the button shall trigger the satellite without any other automation. Did you use my version of the satellite too?

I did:

`~/wyoming-satellite/wyoming_satellite $ mv satellite.py satelliteBACKUP.py
~/wyoming-satellite/wyoming_satellite $ wget https://raw.githubusercontent.com/llluis/wyoming-satellite/remote-detection/wyoming_satellite/satellite.py`

`~/wyoming-satellite $ script/run --name 'Screen Satellite' --uri 'tcp://0.0.0.0:10700' --mic-command 'arecord -D plughw:CARD=BYLM40,DEV=0 -r 16000 -c 1 -f S16_LE -t raw' --snd-command 'aplay -r 22050 -c 1 -f S16_LE -t raw'`

**Comment by llluis** [comment]
> Do you mind sharing your proof of concept so that I may implement and take a look at?

You must use my custom version of the Wyoming protocol in Home Assistant available here:
https://github.com/llluis/homeassistant-wyoming/tree/dev

And my custom version of the satellite available here:
https://github.com/llluis/wyoming-satellite/tree/dev

This will enable:
- Button Trigger detection: the same as speaking the wake work. Will run the whole pipeline
- Button Ask: it will only trigger the speech to text in the satellite to capture an answer from user. No intent matching nor reply / text to speech will be run
- Satellite media_player: bonus to enable text to speech in the satellite without using other software like MPD.
- Pipeline will trigger home assistant events with name `wyoming-satellite-pipeline`

![image](https://github.com/rhasspy/wyoming-satellite/assets/708868/b2f6c568-ecfb-4089-a207-542ff567daff)

So, you would, in an automation:
- Send an TTS to the media_player asking a question
- Wait x seconds to allow the player to ask the question in the speaker
- Trigger the button ask
- Capture the event `wyoming-satellite-pipeline` and filter the type `stt-end`. The field `stt-output/text` will contain your reply. 
![image](https://github.com/rhasspy/wyoming-satellite/assets/708868/fdcdd1e0-2715-409e-97c8-6245ca2ce282)

**Comment by llluis** [comment]
Just to let you know I pushed a few changes to make it easier in the wait_for_trigger automation.
(just download again from the repo if already downloaded before)

![image](https://github.com/rhasspy/wyoming-satellite/assets/708868/470d20a5-7d78-4895-a9af-c0ca63f8ecdc)

**Comment by llluis** [comment]
To wait the reply:

```
wait_for_trigger:
  - platform: event
    event_type: wyoming-satellite-pipeline
    event_data:
      satellite_name: Nabu1
      type: stt-end
```

The reply will be in `wait.trigger.event.data.data.stt_output.text`

**Comment by geman220** [comment]
So `developer-tools/event`

I listen to `wyoming-satellite-pipeline`

If I press `Trigger ask question` I see nothing in the event log.

Restarting the satellite I do see `run-start` and `wake_word-start`, if I use the wake word I see normal `stt-start` and `stt-end`.  But pressing the buttons in the UI doesn't log anything when listening to `wyoming-satellite-pipeline`

**Comment by llluis** [comment]
I must be forgotting something in the setup as it works here. You are seeing the button, so the deploy of the home assistant part went ok. You did copy all files in the wyoming directory, right?

![image](https://github.com/rhasspy/wyoming-satellite/assets/708868/024a0915-9a26-4612-8ccb-53ba0436c520)

Can you post the debug log of the satellite?

Here's the expected output when clicking the ask button:
```
Feb 06 15:23:38 nabu1 run[17414]: DEBUG    Detection called. Name: ask
Feb 06 15:23:38 nabu1 run[17414]: DEBUG    Detection triggered from event
Feb 06 15:23:38 nabu1 run[17414]: DEBUG    Streaming audio
Feb 06 15:23:38 nabu1 run[17414]: DEBUG    RunPipeline from PipelineStage.ASR to PipelineStage.ASR
```

**Comment by geman220** [comment]
I downloaded the .zip from here [https://github.com/llluis/homeassistant-wyoming/tree/dev](https://github.com/llluis/homeassistant-wyoming/tree/dev).  I deleted my old wyoming folder in `\config\custom_components`, created a new wyoming folder `\config\custom_components\wyoming`, and then placed all of the files from the .zip.  I restarted HA after.  I saw the new buttons so all went well there.

I did git pull your dev branch [https://github.com/llluis/wyoming-satellite/tree/dev](https://github.com/llluis/wyoming-satellite/tree/dev) and just to be sure I checked for the recent changes in `satellite.py` and they were there.

I just went to get you a debug log and noticed it's working now.  I haven't changed anything as far as your wyoming addon for HA or satellite dev.  I'm adding the rest of this for context in case someone else may find it useful.

That being said,  I did change my `Whisper` to a docker container on a different server so I could leverage the GPU to increase response speed.  I'm not sure how or why that would make any difference in this case, but the whisper processing was around 15-20 seconds before and now it's down to 1-2s.  I also moved `OpenWakeWord` to the Pi to do local detection of `wakeword`.  So potentially some change in that stack resolved the issue?

I'm going to work on creating an automation now and I will report back.

Thanks!

**Comment by llluis** [comment]
> I also moved OpenWakeWord to the Pi to do **local detection of wakeword**

**This is it**. 
The "hack" I'm using depends on the satellite running on mode local wake word, as I inject the detection the same way the openwakeword does.

**Comment by geman220** [comment]
> > I also moved OpenWakeWord to the Pi to do **local detection of wakeword**
> 
> **This is it**. The "hack" I'm using depends on the satellite running on mode local wake word, as I inject the detection the same way the openwakeword does.

Oh!  Sorry, I didn't realize that was required or I would have set that up sooner!

**Comment by geman220** [comment]
I have it working in node-red, I need to tie some things together to make it fully functioning, but essentially the following works:

Inject a timestamp which fires a TTS "Do you want to turn the lights on?" delay 2 seconds, press the "ask question" button and get the SST.  I listen to the satellite-pipeline and detect `yes.` or `no.` and process automation with a switch.  In my screenshot `Debug 5` has a payload of "yes.".  So I can just change the trigger (timestamp injection) into something useful, and the execution of automation after the yes/no switch.

![image](https://github.com/rhasspy/wyoming-satellite/assets/17059460/3ccf350b-ca04-4fc2-84d7-5831ea5f6285)

**Comment by geman220** [comment]
As an update, I added a more useful trigger and execution.  Works great.

![image](https://github.com/rhasspy/wyoming-satellite/assets/17059460/85a44a68-2eca-4582-ab11-f5e0d9477851)

**Comment by llluis** [comment]
Are you using the media_player from Wyoming or MPD?

**Comment by geman220** [comment]
> Are you using the media_player from Wyoming or MPD?

I am using your implementation of media_player, I figure that would be more useful to use so I could validate everything works as implemented.

**Comment by llluis** [comment]
@geman220, if you update both components, I now send a new event from satellite to home assistant after the TTS ended playing:

![image](https://github.com/rhasspy/wyoming-satellite/assets/708868/60d1654c-62a3-433e-9025-24a6a94f7c39)

This way you could replace the time delay by this event, so it would work for any TTS text lenght automatically.

**Comment by geman220** [comment]
> @geman220, if you update both components, I now send a new event from satellite to home assistant after the TTS ended playing:
> 
> ![image](https://private-user-images.githubusercontent.com/708868/303229575-60d1654c-62a3-433e-9025-24a6a94f7c39.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDczNzMyMTIsIm5iZiI6MTcwNzM3MjkxMiwicGF0aCI6Ii83MDg4NjgvMzAzMjI5NTc1LTYwZDE2NTRjLTYyYTMtNDMzZS05MDI1LTI0YTZhOTRmN2MzOS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMjA4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDIwOFQwNjE1MTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mMGMwMDdjZGI2ZGM4NzUwNjMwZjA2MmVkOWZjZTdmNjYzMzQ2ZGZiZmY3ZTRhODQ3Y2Y0NTRiZDBhN2Y1Nzk2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9._iFtBEZKES-RdEXM9w10NLSkECP32LRebga-7fRRlOk)
> 
> This way you could replace the time delay and wait for this event, so it would work for any TTS text lenght automatically.

You’re a legend. I was actually trying to do a hacky work around for this so I don’t have to guess the wait length.  I will get it updated tomorrow and report back.

**Comment by geman220** [comment]
Just to verify, I did a `git pull` of [https://github.com/llluis/wyoming-satellite/tree/dev](https://github.com/llluis/wyoming-satellite/tree/dev) and replaced the custom component with [https://github.com/llluis/homeassistant-wyoming/tree/dev](https://github.com/llluis/homeassistant-wyoming/tree/dev).

In Node-Red, using a timestamp to inject tts.speak to tts.piper `{"message": "Help computer.","media_player_entity_id":"media_player.screen_satellite_satellite_speaker"}` I don't see any events when listening to `wyoming-satellite-pipeline`.

**Comment by llluis** [comment]
> Just to verify, I did a `git pull` of https://github.com/llluis/wyoming-satellite/tree/dev and replaced the custom component with https://github.com/llluis/homeassistant-wyoming/tree/dev.
> 
> In Node-Red, using a timestamp to inject tts.speak to tts.piper `{"message": "Help computer.","media_player_entity_id":"media_player.screen_satellite_satellite_speaker"}` I don't see any events when listening to `wyoming-satellite-pipeline`.

I named the event differently as it's an event from the satellite and not a "pipeline" event like the others. Check the screenshot. So the name is just `wyoming-satellite`.

**Comment by geman220** [comment]
> > Just to verify, I did a `git pull` of https://github.com/llluis/wyoming-satellite/tree/dev and replaced the custom component with https://github.com/llluis/homeassistant-wyoming/tree/dev.
> > In Node-Red, using a timestamp to inject tts.speak to tts.piper `{"message": "Help computer.","media_player_entity_id":"media_player.screen_satellite_satellite_speaker"}` I don't see any events when listening to `wyoming-satellite-pipeline`.
> 
> I named the event differently as it's an event from the satellite and not a "pipeline" event like the others. Check the screenshot. So the name is just `wyoming-satellite`.

Same result, listening to `wyoming-satellite` and using timestamp to fire a TTS doesn't show any events.

**Comment by evgenyyy** [comment]
I'm also trying this out as I want to play responses from Pi on a different speaker and need to capture TTS response. The original satellite.py that you linked above is not there anymore, so I tried replacing it with the satellite.py here: https://github.com/llluis/wyoming-satellite/blob/master/wyoming_satellite/satellite.py

As soon as I do this and reboot the Pi, wyoming-satellite.service fails to start

**Comment by llluis** [comment]
> I'm also trying this out as I want to play responses from Pi on a different speaker and need to capture TTS response. The original satellite.py that you linked above is not there anymore, so I tried replacing it with the satellite.py here: https://github.com/llluis/wyoming-satellite/blob/master/wyoming_satellite/satellite.py
> 
> As soon as I do this and reboot the Pi, wyoming-satellite.service fails to start

The master branch will not work. You have to point to the dev branch.

**Comment by llluis** [comment]
Sorry folks. There`s a commit missing in the HA part. I'll get this published at night when I'll be back home.

**Comment by evgenyyy** [comment]
> > I'm also trying this out as I want to play responses from Pi on a different speaker and need to capture TTS response. The original satellite.py that you linked above is not there anymore, so I tried replacing it with the satellite.py here: https://github.com/llluis/wyoming-satellite/blob/master/wyoming_satellite/satellite.py
> > As soon as I do this and reboot the Pi, wyoming-satellite.service fails to start
> 
> The master branch will not work. You have to point to the dev branch.

Silly me, confirming the dev branch satellite.py works and I can now trigger detection with a button. Will test the event listening once you've done the commit.

**Comment by llluis** [comment]
Hello! Commit published. Please update the file and try again. :)

**Comment by geman220** [comment]
![image](https://github.com/rhasspy/wyoming-satellite/assets/17059460/c5352b7d-d579-41d9-9396-529e282d6842)

Works now.  I'll have to work on implementing this into node-red so that it waits for this event before starting to listen.

I'm also working on this.  This will send a custom TTS when the assistant "wakes up" You can save this as `awake.sh` then, in your satellite service (or however you are running it) add `--detection-command '/home/user/location/to/awake.sh'`

awake.sh
`#!/bin/bash
echo $(date) "[awake.sh] ...Starting awake.sh script"

# Home Assistant Server Address
HA_SERVER="YOURSERVER"

# Home Assistant Long-Lived Access Token
HA_TOKEN="YOURTOKEN"

# TTS service endpoint
TTS_SERVICE_ENDPOINT="/api/services/tts/speak"

# Entity ID of the media player and TTS entity
MEDIA_PLAYER_ENTITY_ID="media_player.YOURMEDIAPLAYER"
TTS_ENTITY_ID="tts.piper"

# Message to be spoken
TTS_MESSAGE="How can I help?"

# Cache setting
TTS_CACHE="false"

# JSON payload with flat structure
PAYLOAD=$(cat <<EOF
{
  "media_player_entity_id": "$MEDIA_PLAYER_ENTITY_ID",
  "message": "$TTS_MESSAGE",
  "cache": $TTS_CACHE,
  "entity_id": "$TTS_ENTITY_ID"
}
EOF
)

# Send the TTS command to Home Assistant
curl -X POST -H "Authorization: Bearer $HA_TOKEN" -H "Content-Type: application/json" -d "$PAYLOAD" "$HA_SERVER$TTS_SERVICE_ENDPOINT"

echo $(date) "[awake.sh] ...TTS command sent"`

The challenge here is that, in this use case, it will say `do you want me to dim the lights up` then `how can I help` since that's the `--detection-command`.  So I am going to try to implement logic to bypass `--detection-command` if the wakeword is bypassed.

**Comment by llluis** [comment]
> The challenge here is that, in this use case, it will say `do you want me to dim the lights up` then `how can I help` since that's the `--detection-command`. So I am going to try to implement logic to bypass `--detection-command` if the wakeword is bypassed.

Easiest way would be to create a new event in wyoming protocol to use when the wakeword is bypassed. I talked about it with synesthesiam. That would also open other doors with controlling the pipeline execution. For now I would like to focus on getting these new features tested and integrated in the main branch before going too far in the side track. :)
But I'll try to think about it...

**Comment by geman220** [comment]
![image](https://github.com/rhasspy/wyoming-satellite/assets/17059460/b2a0d111-2ab4-4af9-8a8b-944237ee83d8)

Works as expected.  I changed the node `Ask to dim lights up` to a very long sentence, it didn't call `button.press` until it was finished reading the long message.  So it is successfully awaiting the TTS to complete dynamically.  Now I need to figure out how to have it delay itself so it doesn't hear "How can I help?" as set by `--detection-command`

**Comment by llluis** [comment]
I published new features in my branches. Including this one:

![image](https://github.com/rhasspy/wyoming-satellite/assets/708868/229bb80f-8aa9-48f1-89d8-354acb6bacb7)

However it's a breaking change: I removed both buttons in the interface and replaced by a service. It's much more clean as it's use is always in an automation. 

I'll create a doc page to publish those changes and avoid linking all issues to this one...

**Comment by geman220** [comment]
> I published new features in my branches. Including this one:
> 
> ![image](https://private-user-images.githubusercontent.com/708868/304216122-229bb80f-8aa9-48f1-89d8-354acb6bacb7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDc3NzA3NjgsIm5iZiI6MTcwNzc3MDQ2OCwicGF0aCI6Ii83MDg4NjgvMzA0MjE2MTIyLTIyOWJiODBmLThhYTktNDhmMS04OWQ4LTM1NGFjYjZiYWNiNy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMjEyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDIxMlQyMDQxMDhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lZWU2MmJkODJiYzk3ZTkyMDNjZWYxYjliMDcwOWUyNzhhNDc5OTZkMjIyNDUyMzJjY2E0NTExYTk4NjhlOTQ2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.xkWHfex8Fi3JbyRiqKXpQdh5I3leLCXsO_nxGzy-BCA)
> 
> However it's a breaking change: I removed both buttons in the interface and replaced by a service. It's much more clean as it's use is always in an automation.
> 
> I'll create a doc page to publish those changes and avoid linking all issues to this one...

So if I understand correctly, you're making the "listen" trigger by a service call instead of a `button.press`?

**Comment by llluis** [comment]
> > I published new features in my branches. Including this one:
> > ![image](https://private-user-images.githubusercontent.com/708868/304216122-229bb80f-8aa9-48f1-89d8-354acb6bacb7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDc3NzA3NjgsIm5iZiI6MTcwNzc3MDQ2OCwicGF0aCI6Ii83MDg4NjgvMzA0MjE2MTIyLTIyOWJiODBmLThhYTktNDhmMS04OWQ4LTM1NGFjYjZiYWNiNy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMjEyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDIxMlQyMDQxMDhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lZWU2MmJkODJiYzk3ZTkyMDNjZWYxYjliMDcwOWUyNzhhNDc5OTZkMjIyNDUyMzJjY2E0NTExYTk4NjhlOTQ2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.xkWHfex8Fi3JbyRiqKXpQdh5I3leLCXsO_nxGzy-BCA)
> > However it's a breaking change: I removed both buttons in the interface and replaced by a service. It's much more clean as it's use is always in an automation.
> > I'll create a doc page to publish those changes and avoid linking all issues to this one...
> 
> So if I understand correctly, you're making the "listen" trigger by a service call instead of a `button.press`?

Exactly:

![image](https://github.com/rhasspy/wyoming-satellite/assets/708868/0f7540c3-acbe-4c56-8a7f-b5cece08c099)

**Comment by geman220** [comment]
Sweet, is this an update to both the satellite and the addon, or just the addon?  I can't update and adjust my node-red to validate if you need.

**Comment by llluis** [comment]
Both.

**Comment by geman220** [comment]
Cool, I'll update from the dev branch for the addon, I'm assuming I still only need to replace `satellite.py` on the satellite?

**Comment by llluis** [comment]
Awesome, thank you! Yup, just it.

**Comment by geman220** [comment]
Works great.  It is, essentially, a drop-in replacement for the previous implementation.  You just need to change the entity to send the TTS, and call the service `wyoming.remote_trigger` on the satellite entity, instead of `button.press`.

**Comment by llluis** [comment]
Exactly. And you need to set the Question ID to whatever you like so the satellite stops after the STT without trying to match an intent and send a reply. The plan is to use that ID in the future to match with the question but that's not in place yet.

And you probably noted I changed the name of the speaker. Just trying to improve step by step. :)

**Comment by geman220** [comment]
I did notice the speaker name change, I didn't change or add a `question_id`.  I am still listening to the `wyoming-pipeline` from the node-red flow `askToDimUp`.  Which is how it knows what question is being asked and how to handle the response.  Are you saying I can just set a `question_id` on the call service for `wyoming.remote_trigger`?

**Comment by llluis** [comment]
Hello folks, thanks for your support.
Just to be sure the understanding of the remote trigger:

- If you don't supply a question ID, the satellite will behave exactly as if you have said the wake word. So, it's just a bypass. When triggered, the satellite will start listening for a _command_, i.e., you will have to ask for something, the satellite will listen and then try to match the intent and provide a reply via voice. You can always capture those events in the server via `wyoming-satellite-pipeline` but cannot prevent the satellite from running the whole cycle.

- If you do supply a question ID however, the satellite will also start to listen, but you don't have to give a command as it will not try to match an intent. It will just listen, perform STT, transcribe and send the event. There will be no action and no reply. You can capture the reply via the same event `wyoming-satellite-pipeline`.

So, when using in an automation to ask something to the user, you would supply the question_id. 
For instance, trigger a TTS asking if you want to turn on the lights. Fire the event with a question_id, capture the reply and verify if you said YES or something else. If yes, turn on the lights.

Right now, you can put anything in the question_id but in the future I plan to include it in the event so you could match a reply with an specific question.

**Comment by llluis** [comment]
> I guess we will still need to maintain the conversation_id so we can have a more human type of interaction, not only a simple question and answer type of conversation.

For sure. As those requires more deep changes in other areas (home assistant, pipeline, conversation, wyoming itself), I'm leaving it to synesthesiam. He mentionned in another ticket he would include that.

**Comment by Chreece** [comment]
Hello!
Great work done @llluis you solved many problems I had with the original implementetion...
It's been a while and many things changed since februar.
I'm using your satellite + wyoming code with docker also running openwakeword locally for the satellite. Unfortunately as satellite_name: I'm getting the: 'openwakeword', any clues?

**Comment by Chreece** [comment]
> Regarding the bug, a correction was pushed in the upstream code. I need to get those merged in my branch.

Fingers crossed...

Is there any way to test the code?

**Comment by joshnewman17** [comment]
I was able to get this almost completely setup but unfortunately for some reason even when I supply a question ID it still responds to my "yes" response. Is there some sort of requirement for how the question ID should be formatted?

**Comment by KotBayun69** [comment]
This is awesome!
Thank You very mutch for this!!!

I was able to create an automation which asks me if I want to rurn the light on when the room is getting dark and it waits for my response. If I say 'Yes'  it turns the light on, if 'No' or there was no response it does nothing.

Also I wasn't so happy with a standard response from Assist if no intent was recognized and to you should say the wake word again to start over. With Your code I created an atomation which listens to error event and triggers the satellite to listen the command without saying the wake word. Also the TTS now is playing on my snapserver. Great!

**Comment by tinyfrenchcars** [comment]
Hi guys ! 
I'm really interested in that work and wish to give it a try, but I wonder what the state of development, because a few versions of HA are out since the 2024.1.5, and wonder if it will work with the latest. I tried but was a bit confused as the wyoming addon is "integrated" now (unless I misunderstood something), especially on the replacement of the wyoming addon. If any help is possible I'd be gratefull :-)

**Comment by chrispazz** [comment]
Great work! Tried to remove Alexa Actionable notifications.
Can't wait that this will be merged on master branch.

Any way to activate satellite leds when using the remote trigger? currently they do not lights up....

**Comment by KotBayun69** [comment]
> Any way to activate satellite leds when using the remote trigger? currently they do not lights up....

For me remote trigger works as if it was triggered by the wake word.

**Comment by chrispazz** [comment]
> > Any way to activate satellite leds when using the remote trigger? currently they do not lights up....
> 
> For me remote trigger works as if it was triggered by the wake word.

Yes, my mistake! now it is working... :)

**Comment by chrispazz** [comment]
> ![image](https://private-user-images.githubusercontent.com/17059460/303729013-b2a0d111-2ab4-4af9-8a8b-944237ee83d8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ1OTA5MzEsIm5iZiI6MTcyNDU5MDYzMSwicGF0aCI6Ii8xNzA1OTQ2MC8zMDM3MjkwMTMtYjJhMGQxMTEtMmFiNC00YWY5LThhOGItOTQ0MjM3ZWU4M2Q4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODI1VDEyNTcxMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTUwZWZlMWNiNDIwYTUxY2FhNzhhMDBkNDIyMmRjMmEwYzNhYTAwOTk4M2I0YzRlZGQ4NWUwYmIwYWEyMDUyNjAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.m4FmK72dsNBIq2o5bu_Ko_7KXz_WOjbjvx2C7B3nQY8)
> 
> Works as expected. I changed the node `Ask to dim lights up` to a very long sentence, it didn't call `button.press` until it was finished reading the long message. So it is successfully awaiting the TTS to complete dynamically. Now I need to figure out how to have it delay itself so it doesn't hear "How can I help?" as set by `--detection-command`

Hi, can you provide json of your flow? Thank you :)

**Comment by ThreepE0** [comment]
Is this branch still (hopefully) going to be merged in?  I'd like to try this out, but being 9 months without a commit, I'm concerned it's too far behind

**Comment by markselias** [comment]
I would also be very interested in getting this into the main branch! Would also be happy to help with integrating it. @llluis are you still using this? Any plans on developing this further?
@synesthesiam would a PR of this feature into main be interesting or will this not happen anytime soon?

**Comment by crissmil** [comment]
I'm also using this since several months and I love it, please merge

**Comment by KotBayun69** [comment]
Using the custom version of Wyoming Satellite on the specified hardware generally yields good results. However, there are two recurring issues that impact the overall stability and usability of the system.

- **Mic Service Disconnects:**

    _Description_: After running the satellite several times in a row, the microphone service unexpectedly disconnects. This requires manual intervention to restore functionality.
    _Frequency_: Occurs intermittently, typically after 4–6 successive runs.

- **Service Reboots:**

    _Description_: The enhanced-wyoming-satellite.service periodically restarts on its own without any clear trigger.
    _Frequency_: Happens sporadically, but at least once during extended usage sessions.

**Hardware Setup:**

Raspberry Pi Zero 2W
ReSpeaker 6 Mic Array
